plutus-tx/src/PlutusTx/Data/AssocList.hs

```diff
@@ -0,0 +1,436 @@
+-- In case of duplicates, this function will keep only one entry (the one that precedes).
+-- In other words, this function de-duplicates the input list.
+safeFromList :: forall k a . (Eq k, P.ToData k, P.ToData a) => [(k, a)] -> AssocList k a
+safeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+    . foldr (uncurry go) []
+  where
+    go :: k -> a -> [(k, a)] -> [(k, a)]
+    go k v [] = [(k,  v)]
+    go k v ((k', v') : rest) =
+      if k == k'
+        then (k, v) : rest
+        else (k', v') : go k v rest
+
+{-# INLINEABLE unsafeFromList #-}
+-- | Unsafely create an 'AssocList' from a list of pairs.
+-- This should _only_ be applied to lists which have been checked to not
+-- contain duplicate keys, otherwise the resulting 'AssocList' will contain
+-- conflicting entries (two entries sharing the same key), and therefore be ill-defined.
+unsafeFromList :: (P.ToData k, P.ToData a) => [(k, a)] -> AssocList k a
+unsafeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+
+{-# INLINEABLE uncons #-}
+-- | Decompose an 'AssocList' into its first key-value pair and the rest of the list.
+uncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  AssocList k a ->
+  Maybe ((k, a), AssocList k a)
+uncons m = case P.uncons (toBuiltinList m) of
+  Nothing -> Nothing
+  Just (pair, rest) ->
+    let (k, a) = P.pairToPair pair
+     in Just ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList rest)
+
+{-# INLINEABLE unsafeUncons #-}
+-- | Decompose an 'AssocList' into its first key-value pair and the rest of the list.
+-- This function is unsafe because it assumes that the `AssocList` is not empty.
+unsafeUncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  AssocList k a ->
+  ((k, a), AssocList k a)
+unsafeUncons m =
+  ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList tl)
+  where
+    (hd, tl) = P.unsafeUncons (toBuiltinList m)
+    (k, a) = P.pairToPair hd
+
+{-# INLINEABLE noDuplicateKeys #-}
+-- | Check if the `AssocList` is well-defined. Warning: this operation is O(n^2).
+noDuplicateKeys :: forall k a. AssocList k a -> Bool
+noDuplicateKeys m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let k = BI.fst hd
+             in if member' k tl then False else go tl
+        )
+
+{-# INLINEABLE all #-}
+--- | Check if all values in the `AssocList` satisfy the predicate.
+all :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> AssocList k a -> Bool
+all p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
```

Comment:
Also, if `matchList` took the `xs` argument last, we could eta-contract `go` as well.